src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+                this.regexIndexer.forEachMatchWithinRange(this.regexIndexer, startPosition, endPosition, fnResult);
+            },
+
+            /**
+             * Gets the start and end positions plus the regular expression match array data
+             * @return {{to: {line: number, ch: number}, from: {line: number, ch: number}, match: Array}} returns start and end of match with the array of results
+             */
+            getFullInfoForCurrentMatch: function () {
+                var docText = _getDocumentText(this.doc);
+                return this.regexIndexer.getFullResultInfo(this.regexIndexer.getCurrentMatchNumber(), this.query, docText);
+            },
+
+            /**
+             * Creates an Array of integers which is filled with a pattern matching the lines
+             * of the document which contain matches.
+             * A value of '0' is no match.
+             * A value of '1' is positive match.
+             *
+             * @param   {[[Type]]} levelOfDetail An Array of this size will be created and filled with a pattern matching lines matched
+             * @returns {object}   An array filled with line match pattern
+             */
+            createMatchedLinePattern: function (levelOfDetail) {
+                _updateResultsIfNeeded(this);
+                var representationArray = new Uint8Array(levelOfDetail);
+                var docLineIndex = _getDocumentIndex(this.doc);
+                var linesPerArraySlot = docLineIndex.length / levelOfDetail;
+                this.regexIndexer.fillWithMatchedLinePattern(representationArray);
+                return {linesPerArraySlot: linesPerArraySlot, lineMatchPatternArray: representationArray};
+            },
+
+            /**
+             * Finds the indexes of all matches based on the current search query
+             * The matches can then be navigated and retrieved using the functions of the search cursor.
+             *
+             * @return {number} the count of matches found.
+             */
+            scanDocumentAndStoreResultsInCursor: function () {
+                if (_needToIndexDocument(this.doc)) {
+                    _indexDocument(this.doc);
+                }
+                var docText = _getDocumentText(this.doc);
+                var docLineIndex = _getDocumentIndex(this.doc);
+                this.regexIndexer = _createRegexIndexer(docText, docLineIndex, this.query, this.maxResults, this.currentPosition);
+                this.resultsCurrent = true;
+                return this.getMatchCount();
+            }
+        });
+    }
+
+    /**
+     * Creates an updatable search cursor which can be used to navigate forward and backward through the results.
+     *
+     * @param {!{document: CodeMirror.Doc, searchQuery: string|RegExp, position: {line: number, ch: number}, ignoreCase: boolean}} properties
+     * @return {Object} The search cursor object
+     */
+    function createSearchCursor(properties) {
```

Comment:
Why passing `properties` instead of named arguments? It's understandable while in development (while exact list of arguments is not defined), but not in its final form.